authui/src/web3.ts

```diff
@@ -207,17 +220,31 @@ export class WalletConfirmationController extends Controller {
   declare submitTarget: HTMLButtonElement;
 
   declare providerValue: string;
-  declare provider: ethers.providers.Web3Provider | null;
+  declare provider: Web3Provider | null;
 
   connect() {
-    this.provider = getProvider(this.providerValue);
+    getProvider(this.providerValue)
+      .then((provider) => {
+        if (!provider) {
+          visit(`/missing_web3_wallet?provider=${this.providerValue}`);
+          return;
+        }
+        this.provider = provider;
+        this._getAccount();
+      })
+      .catch((err) => {
+        handleError(err);
+      });
 
-    if (!this.provider) {
-      visit(`/missing_web3_wallet?provider=${this.providerValue}`);
-      return;
+    if (isProviderMetaMask(this.provider?.provider)) {
+      this.provider!.provider.on("accountsChanged", () => this._getAccount());
     }
+  }
 
-    this._getAccount();
+  disconnect() {
+    if (isProviderMetaMask(this.provider?.provider)) {
+      this.provider!.provider.off("accountsChanged", () => this._getAccount());
```

Comment:
The on/off pair MUST be passed the same function. Here we always pass them a fresh arrow function. So the function is leaked.

Here is how to accommodate the on/off pattern correctly.
```
class Foobar {
  connect() {
    this.provider.provider.on("accountsChanged", this.onAccountsChanged);
  }
  disconnect() {
    this.provider.provider.off("accountsChanged", this.onAccountsChanged);
  }

  // This is very important to define onAccountsChanged as a property.
  // It is bound to this.
  // If it were defined as
  // async onAccountsChanged() { }
  // We need to bind it with this.onAccountsChanged.bind(this)
  // But .bind always return a fresh bound function,
  // so our problem will remain unfixed.
  onAccountsChanged = async () => {
  }
}
```